flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/converters/SqlNodeConvertUtils.java

```diff
@@ -63,6 +69,32 @@ static CatalogView toCatalogView(
         // This bug is fixed in CALCITE-3877 of Calcite 1.23.0.
         String originalQuery = context.toQuotedSqlString(query);
         SqlNode validateQuery = context.getSqlValidator().validate(query);
+
+        // Check name is unique.
+        // Don't rely on the calcite because if the field names are duplicate, calcite will add
+        // index to identify the duplicate names.
+        SqlValidatorNamespace namespace = context.getSqlValidator().getNamespace(validateQuery);
+        Map<String, Integer> nameToPos = new HashMap<>();
+        for (int i = 0;
+                i < Objects.requireNonNull(namespace).getType().getFieldList().size();
+                i++) {
+            String columnName = namespace.getType().getFieldList().get(i).getName();
+            if (nameToPos.containsKey(columnName)) {
+                SqlSelect select = (SqlSelect) query;
```

Comment:
It may be a SqlCall, e.g. Union or CTE. Please add more test cases here to verify the behaviour.